reflect.go

```diff
@@ -0,0 +1,456 @@
+			notCheckedAPIs[knownAPI] = true
+		}
+	}
+
+	tf.ignoreReflectedTypes(ssaPkg)
+
+	// all previously unchecked APIs have now been checked add them to checkedAPIs,
+	// to avoid checking them twice
+	maps.Copy(checkedAPIs, notCheckedAPIs)
+
+	// if a new reflectAPI is found we need to Re-evaluate all functions which might be using that API
+	if len(cachedOutput.KnownReflectAPIs) > lenPrevKnownReflectAPIs {
+		tf.recordReflection(ssaPkg)
+	}
+}
+
+// find all functions, methods and interface declarations of a package and record their
+// reflection use
+func (tf *transformer) ignoreReflectedTypes(ssaPkg *ssa.Package) {
+	for _, memb := range ssaPkg.Members {
+		switch x := memb.(type) {
+		case *ssa.Type:
+			// methods aren't package members only their reciever types are
+			// so some logic is required to find the methods a type has
+
+			method := func(mset *types.MethodSet) {
+				for i, n := 0, mset.Len(); i < n; i++ {
+					at := mset.At(i)
+
+					if f := ssaPkg.Prog.MethodValue(at); f != nil {
+						tf.checkCalls(nil, f)
+					} else {
+						fun := at.Obj().(*types.Func)
+						// handle interface declarations
+						tf.checkCalls(fun, nil)
+					}
+
+				}
+			}
+
+			// yes, finding all methods really only works with both calls
+			mset := ssaPkg.Prog.MethodSets.MethodSet(x.Type())
+			method(mset)
+
+			mset = ssaPkg.Prog.MethodSets.MethodSet(types.NewPointer(x.Type()))
+			method(mset)
+
+		case *ssa.Function:
+			// these not only include top level functions, but also synthetic
+			// functions like the initialization of global variables
+
+			tf.checkCalls(nil, x)
+		}
+	}
+}
+
+// Exported methods with unnamed structs as paramters may be "used" in interface declarations
+// elsewhere, these interfaces will break if any method uses reflection on the same parameter.
+//
+// Therefore never obfuscate unnamed structs which are used as a method parameter
+// and treat them like a parameter which is actually used in reflection.
+//
+// See "UnnamedStructMethod" in the reflect.txtar test for an example.
+func (tf *transformer) checkMethodSignature(reflectParams map[int]bool, sig *types.Signature) {
+	if sig.Recv() == nil {
+		return
+	}
+
+	params := sig.Params()
+	for i := 0; i < params.Len(); i++ {
+		if reflectParams[i] {
+			continue
+		}
+
+		ignore := false
+		param := params.At(i)
+
+		switch x := param.Type().(type) {
+		case *types.Struct:
+			ignore = true
+		case *types.Array:
+			if _, ok := x.Elem().(*types.Struct); ok {
+				ignore = true
+			}
+		case *types.Slice:
+			if _, ok := x.Elem().(*types.Struct); ok {
+				ignore = true
+			}
+		}
+
+		if ignore {
+			reflectParams[i] = true
+			tf.recursivelyRecordAsNotObfuscated(param.Type())
+		}
+	}
+}
+
+// Checks all callsites in a function declaration for use of reflection.
+func (tf *transformer) checkCalls(f *types.Func, fun *ssa.Function) {
+	if f != nil && fun != nil {
+		panic("either provide a *types.Func or a *ssa.Function not both")
+	}
```

Comment:
you always call checkCalls with either a non-nil f, or a non-nil fun, but not both. why not split this into two funcs? you would only share four lines - the `make` of the map, and the `if len(reflectParams) > 0 {}` at the end of the func. The rest of the code could get unindented too, e.g. `if fun != nil {` can go away.

Reply:
when you do split them, you can also call them more helpful names, like `checkFunction` for `fun`, and `checkInterfaceMethod` for `f`.